### PR TITLE
Use half the core count on failing CI platforms

### PR DESCRIPTION
**Pull request by eddelbuettel**
PR #4182 saw CI fail for the first few commits, with apparent resource exhaustion for Linux/debug and macOS/cli.  After adjusting the number of cores, all subsequent attempts passed: three resulted in 'green', one ended on an unrelated (markdown) issue but well after the earlier failures due to resource exhaustion.

### CODE REVIEW

**Comment by rcurtin** [comment]
This will certainly slow down CI, which doesn't bother me so much, but I know is a priority of yours.  There may be another way out here: the file that caused the RAM exhaustion, `layer_serialization_test.cpp`, is actually only testing `BatchNorm` layer serialization for the neural network code.

A bit of background: in order to be able to serialize layers, because it is a class hierarchy with inheritance, you have to explicitly register those layers with cereal.  So the [serialization.hpp file](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/serialization.hpp) registers every known layer with cereal.

But this has the drawback that the compiler now must instantiate every method for every layer, and this can itself take quite a lot of time and resources.

So, one possibility is to ditch the `#define MLPACK_ENABLE_ANN_SERIALIZATION` and replace it with `CEREAL_REGISTER_TYPE(mlpack::BatchNorm<arma::mat>);`.  You may also need to `#define MLPACK_ANN_IGNORE_SERIALIZATION_WARNING` (see [here](https://github.com/mlpack/mlpack/blob/c0d891a836cb7e16fde3458fff82c5baf503ea0c/src/mlpack/methods/ann/ffn_impl.hpp#L383)) if an `FFN` or `RNN` object is being serialized (I don't remember what the test says).

I suspect that enabliing serialization only for the `BatchNorm` layer would significantly reduce resource usage overall.  If more serialization tests were added to that file, we'd probably want to just add the `CEREAL_REGISTER_TYPE()` macro one by one.

Anyway, let me know what you think: I am fine with either solution.

**Comment by eddelbuettel** [comment]
> This will certainly slow down CI,

That's why my commit message in the other PR started with _Speculative: ..._  And I watched four runs, and it did not worsen things there.  Which may be a function of what code I altered / what `ccache` can do for us.  But I think this change is fine, and that is why I made it a PR.

Extending `cmake` logic to actually be more reasonable with resource use is also a fine idea, but not something I will pursue here.  We may end up needing it sooner or later though as our code grows while the CI resources remain fixed ...

**Review by rcurtin — APPROVED** [review]
Agree, and if you didn't see a slowdown in CI then hey, maybe it is less painful than I thought.

Anyway, I do know the serialization idea I wrote will help, but we can save that for another day if the problem is solved today.

> as our code grows while the CI resources remain fixed ...

I think it might be worse than that!  I think newer compiler versions tend to get more and more resource-hungry, meaning that we lose long-term no matter what.  And I would not be surprised if Github starts cutting back on the resources that runners get in order to save their own costs...

**Comment by eddelbuettel** [comment]
> I think newer compiler versions tend to get more and more resource-hungry,

Yes! I already heard bad stories abouit `gcc-16` !

Another option may be to run CI jobs only when changes affect "their area" (e.g. when the go bindings change, we do not need julia/python/cli/R to run...) though that may be a little tricky here because we often change the underlying C++ layer.  Then again, maybe we do not need cli on multiple platforms, or maybe we can do with debug mode, or ...
But I know you are feeling quite strongly about the status quo so ... wait and see it is.

**Comment by rcurtin** [comment]
Those would be great improvements!  I don't mind changing around how we do CI so long as we aren't losing testing surface.

**Review by zoq — APPROVED** [review]
